server/aws-lsp-codewhisperer/src/language-server/securityScan/securityScanDiagnosticsProvider.ts

```diff
@@ -0,0 +1,136 @@
+import { Logging, Lsp } from '@aws/language-server-runtimes/out/features'
+import { Diagnostic, Hover, Position, Range, TextDocumentContentChangeEvent } from 'vscode-languageserver'
+import { URI } from 'vscode-uri'
+import { AggregatedCodeScanIssue, CodeScanIssue } from './types'
+
+class SecurityScanDiagnosticsProvider {
+    private diagnostics: Map<string, Diagnostic[]>
+    private lsp: Lsp
+    private logging: Logging
+
+    constructor(lsp: Lsp, logging: Logging) {
+        this.diagnostics = new Map()
+        this.lsp = lsp
+        this.logging = logging
+    }
+
+    async createDiagnostics(findings: AggregatedCodeScanIssue[]) {
+        for (const finding of findings) {
+            const path = URI.file(finding.filePath).path
+            const diagnostics = finding.issues.map(issue => this.mapScanIssueToDiagnostics(issue))
+            if (!this.diagnostics.has(path)) {
+                this.diagnostics.set(path, diagnostics)
+            }
+            this.diagnostics.set(path, [...(this.diagnostics.get(finding.filePath) || []), ...diagnostics])
+            await this.publishDiagnostics(finding.filePath, diagnostics)
+        }
+    }
+
+    mapScanIssueToDiagnostics(issue: CodeScanIssue): Diagnostic {
+        return Diagnostic.create(
+            this.createDiagnosticsRange(issue.startLine, issue.endLine),
+            `${issue.title} - ${issue.description.text}`,
+            2,
+            issue.relatedVulnerabilities.join(','),
+            'Detected by CodeWhisperer'
+        )
+    }
+
+    async validateDiagnostics(uri: string, e: TextDocumentContentChangeEvent) {
+        const path = URI.parse(uri).path
+        const currentDiagnostics = this.diagnostics.get(path)
+        if (!currentDiagnostics) {
+            return
+        }
+        // range will be omitted if the change is for entire document
+        const changeStartLine = (e as any).range?.start?.line
```

Comment:
nit: to me this approach(https://github.com/aws/language-servers/pull/157#discussion_r1519751853) still looks cleaner but this is an opinion, not a blocker